### PR TITLE
Add BalancerTokenAdmin contract

### PR DESCRIPTION
**Pull request by TomAFrench**
I've added a rough version of the BalancerInflation contract which is an equivalent to the non-ERC20 components of [ERC20CRV](https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/ERC20CRV.vy).

The main differences are:
- Rather than having a `INFLATION_DELAY` we have an explicit activation function.
  - This results in simplification to `_updateMiningParameters` as we don't need to catch this special case.
- We inherit from `Authentication` to allow the Authorizer to define who can mint rather than defining an explicit minter.
- We expose a `snapshot` function which anyone authorised by that Authorizer can call
- `futureEpochTimeWrite` is implemented as `_startEpochTimeWrite() + _RATE_REDUCTION_TIME` as opposed to [duplicated logic](https://github.com/curvefi/curve-dao-contracts/blob/4e428823c8ae9c0f8a669d796006fade11edb141/contracts/ERC20CRV.vy#L134-L162)

### CODE REVIEW

**Comment by TomAFrench** [comment]
I'm leaving this as draft for a bit pending tests.

File: pkg/gauges/contracts/BalancerTokenAdmin.sol [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+                uint256 currentStart = start;
+                if (currentStart >= currentEpochTime.add(_RATE_REDUCTION_TIME)) {
+                    // We should never get here but what if...
+                    break;
+                } else if (currentStart < currentEpochTime) {
+                    currentStart = currentEpochTime;
+                }
+
+                toMint = toMint.add(currentRate.mul(currentEnd - currentStart));
```
Comment (nventuro) [inline pkg/gauges/contracts/BalancerTokenAdmin.sol:305]:
```suggestion
                toMint = toMint.add(currentRate.mul(currentEnd.sub(currentStart)));
```

I find it so hard to prove that end > start that I'd just add this.

File: pkg/gauges/contracts/BalancerTokenAdmin.sol [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+    /**
+     * @notice Mint BAL tokens subject to the defined inflation schedule
+     * @dev Callable only by addresses defined in the Balancer Authorizer contract
+     */
+    function mint(address to, uint256 amount) external authenticate {
+        require(
+            _balancerToken.totalSupply().add(amount) <= _availableSupply(),
+            "Mint amount exceeds remaining available supply"
+        );
```
Comment (nventuro) [inline pkg/gauges/contracts/BalancerTokenAdmin.sol:168]: CRV has an extra statement here:

```
    if block.timestamp >= self.start_epoch_time + RATE_REDUCTION_TIME:
        self._update_mining_parameters()
```

File: pkg/gauges/contracts/BalancerTokenAdmin.sol [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+     * @param end End of the time interval (timestamp)
+     * @return Tokens mintable from `start` till `end`
+     */
+    function _mintableInTimeframe(uint256 start, uint256 end) internal view returns (uint256) {
+        require(start <= end, "start > end");
+
+        uint256 currentEpochTime = startEpochTime;
+        uint256 currentRate = rate;
+
+        // It shouldn't be possible to over/underflow in here but we add checked maths to be safe
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/gauges/contracts/BalancerTokenAdmin.sol:280]: Those operations are all very safe as we're either incrementing by very small numbers so it's impossible to overflow in a reasonable length of time (i.e. human civilisation will end first) or they converge to values which will not overflow (max supply or zero)

File: pkg/gauges/contracts/BalancerTokenAdmin.sol (hunk 2 above)
Comment (nventuro) [reply pkg/gauges/contracts/BalancerTokenAdmin.sol:174]: The check is also quite cheap for minters that update before mint, right? As `start_epoch_time` will be warm.

File: pkg/gauges/contracts/BalancerTokenAdmin.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/gauges/contracts/BalancerTokenAdmin.sol:315]: These make me a bit anxious as I find it hard to check in all instances that overflow is indeed impossible, and the original source code (the CRV contract) has checked arithmetic everywhere, so it's not like we can rely on hints from them :/

I'll take a detailed look to convince myself of correctness - once that's done, we can merge.

[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/gauges/contracts/BalancerTokenAdmin.sol:315]: I now regret taking a detailed look, but am however convinced of correctness.

**Review by nventuro — APPROVED** [review]

**Pull request merged**